File: canonical/python/src/yx/transport/key_store.py
"""
Per-peer key management.

Traceability:
- specs/architecture/security-architecture.md (Per-Peer Key Management)
"""

from dataclasses import dataclass, field
from typing import Dict
import logging

logger = logging.getLogger(__name__)


@dataclass
class KeyStore:
    """
    Manages per-peer symmetric keys.

    Traceability:
    - specs/architecture/security-architecture.md (KeyStore)
    """
    default_key: bytes  # 32 bytes, fallback key
    _peer_keys: Dict[str, bytes] = field(default_factory=dict)

    def __post_init__(self):
        """Validate default key."""
        if len(self.default_key) != 32:
            raise ValueError("Default key must be 32 bytes")

    def get_key(self, peer_id: str) -> bytes:
        """Get key for peer (or default)."""
        return self._peer_keys.get(peer_id, self.default_key)

    def set_key(self, peer_id: str, key: bytes):
        """Set peer-specific key."""
        if len(key) != 32:
            raise ValueError("Key must be 32 bytes")

        self._peer_keys[peer_id] = key
        logger.info(f"Set key for peer: {peer_id}")

    def remove_key(self, peer_id: str):
        """Remove peer-specific key."""
        if peer_id in self._peer_keys:
            del self._peer_keys[peer_id]
            logger.info(f"Removed key for peer: {peer_id}")

    def has_peer_key(self, peer_id: str) -> bool:
        """Check if peer has specific key."""
        return peer_id in self._peer_keys

    def peer_count(self) -> int:
        """Get number of peers with specific keys."""
        return len(self._peer_keys)

File: swift/builds/swift-impl/Sources/YXProtocol/Transport/KeyStore.swift
import Foundation
import CryptoKit

/// Per-peer symmetric key manager.
///
/// Falls back to defaultKey when no peer-specific key is set.
///
/// Traceability:
/// - protocol/specs/architecture/security-architecture.md (Per-Peer Keys)
actor KeyStore {

    private let defaultKey: Data
    private var peerKeys: [String: Data] = [:]

    /// - Parameter defaultKey: 32-byte fallback key
    init(defaultKey: Data) throws {
        guard defaultKey.count == 32 else {
            throw KeyStoreError.invalidKeySize
        }
        self.defaultKey = defaultKey
    }

    /// Get key for peer (or default).
    func getKey(peerID: String) -> Data {
        return peerKeys[peerID] ?? defaultKey
    }

    /// Set peer-specific key.
    func setKey(_ key: Data, for peerID: String) throws {
        guard key.count == 32 else { throw KeyStoreError.invalidKeySize }
        peerKeys[peerID] = key
    }

    func removeKey(for peerID: String) { peerKeys.removeValue(forKey: peerID) }
    func hasPeerKey(_ peerID: String) -> Bool { peerKeys[peerID] != nil }
    var peerCount: Int { peerKeys.count }
}

enum KeyStoreError: Error, CustomStringConvertible {
    case invalidKeySize
    var description: String { "Key must be 32 bytes for AES-256" }
}
